apps/dashboard/app/apps/ood_app.rb

```diff
@@ -332,4 +341,15 @@ def has_gem?(gemname)
   def bundler_helper
     @bundler_helper ||= BundlerHelper.new(path)
   end
+
+  # add all the metadata fields in manifest.metadata as attr_accessors
+  # to class and set instance variables to this instance.
+  def add_metadata_attrs
+    metadata.reject do |k, _|
+      reserved_metadata_keys.include?('k')
+    end.each do |k,v|
+      self.instance_variable_set("@#{k}", v)
+      self.class.send(:attr_accessor, k)
+    end
+  end
```

Comment:
This method modifies the class. So it should not be executed in initialize, otherwise every time a new instance of OodApp is generated, the class will be modified. But if you take it out of OodApp, then I understand the issue in that every app might have its own subset of metadata keys.

Is the issue here that we have:

```
  def self.groups_for(apps: [], group_by: :category, nav_limit: nil)
    apps.group_by { |app|
      app.try(group_by)
```

So `group_by:` is expected to be a method defined on OodApp so that we can do `apps.group_by { |app| app.try(group_by) }`?

Reply:
This might work, and you may be able to get rid of the sanitizing as a result.

```ruby
  def self.groups_for(apps: [], group_by: :category, nav_limit: nil)
    apps.group_by { |app|
      app.respond_to?(group_by) ? app.send(:group_by) : app.metadata[group_by]
    }
```

Reply:
Yes that works, which I'm a little sour about because I recall trying `respond_to?` in the other PR and settled on `try`.

Either way, yes this works and greatly simplifies the changes required.